src/ml/neural_net/combine_futures_subscriber.hpp

```diff
@@ -0,0 +1,160 @@
+ * be found in the LICENSE.txt file or at
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#ifndef ML_NEURAL_NET_COMBINE_FUTURES_SUBSCRIBER_HPP_
+#define ML_NEURAL_NET_COMBINE_FUTURES_SUBSCRIBER_HPP_
+
+#include <future>
+#include <memory>
+#include <queue>
+
+#include <ml/neural_net/combine_base.hpp>
+
+namespace turi {
+namespace neural_net {
+
+/**
+ * Subscriber that synchronously produces futures for promises to be fulfilled
+ * by its publisher.
+ *
+ * This type is useful for integrating Publishers into existing code bases that
+ * rely on synchronous behavior or futures.
+ *
+ * Client code MUST call FuturesSubscriber::Cancel() to tear down a
+ * FuturesSubscriber instance. This requirement can be handled automatically
+ * using the FuturesStream wrapper class below.
+ */
+template <typename T>
+class FuturesSubscriber : public Subscriber<T> {
+ public:
+  using Input = T;
+
+  /**
+   * Submits a request for a value to the Publisher but immediately returns a
+   * future for that value.
+   *
+   * If the publisher returned a failure for this request or any previous
+   * request from this subscriber, then the future will store that exception. If
+   * the publisher returned Completion::IsFinished() for this request or any
+   * previous request, or if Cancel is called, then the future will store null.
+   *
+   * \todo Use an Optional<T> type instead of unique_ptr to avoid allocation.
+   */
+  std::future<std::unique_ptr<T>> Request() {
+    std::promise<std::unique_ptr<T>> promise;
+    auto future = promise.get_future();
+    if (failure_) {
+      // We've already observed an exception. Set it now.
+      promise.set_exception(failure_);
+    } else if (completed_) {
+      // We've already observed the end of the sequence. Signal completion now.
+      promise.set_value(nullptr);
+    } else {
+      // Enqueue this promise and submit a request to the Publisher.
+      promises_.push(std::move(promise));
+      if (subscription_) {
+        subscription_->Request(Demand(1));
+      }
+    }
+    return future;
+  }
+
+  void Cancel() {
+    if (completed_) return;
+
+    completed_ = true;
+
+    // Cancel the subscription if active.
+    if (subscription_) {
+      subscription_->Cancel();
+      subscription_.reset();
+    }
+
+    // Fulfill any outstanding promises.
+    while (!promises_.empty()) {
+      promises_.front().set_value(nullptr);
+      promises_.pop();
+    }
+  }
+
+  void Receive(std::shared_ptr<Subscription> subscription) override {
+    // Reject any subscriptions after the first. Reject the first subscription
+    // if we cancelled before it could start.
+    if (subscription_ || completed_) {
+      subscription->Cancel();
+      return;
+    }
+
+    subscription_ = std::move(subscription);
+
+    // If we already have promises queued, request their values now.
+    if (!promises_.empty()) {
+      Demand demand(static_cast<int>(promises_.size()));
+      subscription_->Request(demand);
```

Comment:
shouldn't we use the last subscription to do it? I think you want to exhaust the last subscription here.

Reply:
and you also want to clean up `failure_` as well.

Reply:
What do you mean by "last subscription" and "exhaust the subscription"? Typically there's only one subscription between a publisher and subscriber. For example, see rule 12 under "Subscriber" in this open-source specification of the API protocol, which appears to roughly match Apple's Combine: https://github.com/reactive-streams/reactive-streams-swift

Reply:
Also, what is there to clean up for `failure_`? That variable cannot be non-null at this point, since we must receive the subscription before any completion.

Reply:
I mean, the already queued promises correspond to get data from the previous subscription. From here,

```
    if (subscription_ || completed_) {
       subscription->Cancel();
       return;
     }
```

Also `failure_` might be the exception of that canceled subscription, we should reset it for the new subscription passed in.

Reply:
Hmm, let me try to explain my intent for how this class works. As a Subscriber, this class receives from its Publisher a Subscription, zero or more values, then (maybe) a Completion, in that order. Each value corresponds to one value that this class’s user demands via `Request`.

At present, all of this is synchronous, but the class is written to guard against a future case when all of this is asynchronous. In that scenario, the user can call `Request` (and generate a future) _before_ the class has even received a Subscription from its Subscriber. That’s the situation the code here (line 94) is handling. In this case, when we finally receive the Subscription, then we submit the demands to the Publisher to correspond to the requests our user has already made

Reply:
Another way of explaining it: in general, each time `FuturesSubscriber::Request` is called by the user, the implementation submits a Request to its underlying Subscription. But it’s possible that it hasn’t received the Subscription yet.

Reply:
I understand the intent to design for `Request`.

I'm concerned with if I added another subscription mistakenly. As a user, I would expect the subscription is consistent and data I will get is from same subscription. No surprises coming out from a mistakenly switching to other subscriptions.

Also, if I use a new subscription, that means the stream restarts from this publisher and the `failure_` value is from the previous subscription (publisher).

Reply:
Line 85 is intended to handle this case, by rejecting any subscriptions after the first.

(This aligns with Subscriber Rule 5 here: https://github.com/reactive-streams/reactive-streams-swift.)

The intent is to never accept any subscription after the first. It's generally up to the user not to attach one Subscriber to multiple Publishers, but the API design helps here by encouraging the use of operations (like `Map`) that instantiate a Subscriber and then attach to it all at once. In short, by having factory functions on Publisher that both generate the next pipe in the pipeline and connect it immediately, yielding a new Publisher.

Reply:
Yeah, I know the subscription is canceled. I mean the receiver doesn't know this.

The `future` that the subscriber gets corresponds to the `promise` from the canceled subscription. I would expect this way. Otherwise, it's like I subscribe to _New York Times_, and I paid for the annual fee and suddenly I received Washington Posts someday.

I think you could send a failure saying subscription change, renew your subscription to the new publisher.

Reply:
Well, this case is documented in the comments here:
https://github.com/apple/turicreate/blob/8805840d3679fbdc9ab9f59b7df08ecfe2a07826/src/ml/neural_net/combine_futures_subscriber.hpp#L43

This test case covers it: https://github.com/apple/turicreate/blob/15bd9f8ff8b56d511038854a9571028c64fa89ee/test/unity/toolkits/neural_net/test_combine_futures_subscriber.cxx#L87

Since the implementation treats cancellation and receiving a Completion as having the same effect here, setting `completed_` to true

Reply:
Yeah, but this is the case where subscription is completed not due to a subscription change. 

I mean the promises are generated based on the **previous** subscription. We should set those promises with nullptr as well since the **previous** subscription is forced to be finished, which is also consistent with the behavior that a subscription is completed.

Reply:
What previous subscription? By design, we only allow one subscription per subscriber. Can you give a specific code path or test case illustrating your concern?

Reply:
let's discuss that in person later.